pkg/util/subnet/subnetsupdater.go

```diff
@@ -0,0 +1,74 @@
+package subnet
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+import (
+	"strings"
+
+	mgmtnetwork "github.com/Azure/azure-sdk-for-go/services/network/mgmt/2020-08-01/network"
+	"github.com/Azure/go-autorest/autorest/to"
+)
+
+type Updater interface {
+	AddEndpointsToSubnets(endpoints []string, subnets []*mgmtnetwork.Subnet) (updatedSubnets []*mgmtnetwork.Subnet)
+}
+
+type UpdaterManager struct {
+}
+
+func (um *UpdaterManager) AddEndpointsToSubnets(endpoints []string, subnets []*mgmtnetwork.Subnet) (updatedSubnets []*mgmtnetwork.Subnet) {
+	for _, subnet := range subnets {
+		subnetChanged := um.addEndpointsToSubnet(endpoints, subnet)
+		if subnetChanged {
+			updatedSubnets = append(updatedSubnets, subnet)
+		}
+	}
+
+	return updatedSubnets
+}
+
+func (um *UpdaterManager) addEndpointsToSubnet(endpoints []string, subnet *mgmtnetwork.Subnet) (subnetChanged bool) {
+	for _, endpoint := range endpoints {
+		endpointFound, serviceEndpointPtr := um.subnetContainsEndpoint(subnet, endpoint)
+		if !endpointFound {
+			um.addEndpointToSubnet(endpoint, subnet)
+			subnetChanged = true
+			continue
+		}
+
+		if serviceEndpointPtr.ProvisioningState != mgmtnetwork.Succeeded {
+			um.addEndpointToSubnet(endpoint, subnet)
+			subnetChanged = true
+		}
+	}
+
+	return subnetChanged
+}
+
+func (um *UpdaterManager) subnetContainsEndpoint(subnet *mgmtnetwork.Subnet, endpoint string) (endpointFound bool, serviceEndpointPtr *mgmtnetwork.ServiceEndpointPropertiesFormat) {
+	if subnet == nil || subnet.ServiceEndpoints == nil {
+		return false, nil
+	}
+
+	for _, serviceEndpoint := range *subnet.ServiceEndpoints {
+		endpointFound = strings.EqualFold(*serviceEndpoint.Service, endpoint)
+		if endpointFound {
+			return true, &serviceEndpoint
+		}
+	}
+
+	return false, nil
+}
+
+func (um *UpdaterManager) addEndpointToSubnet(endpoint string, subnet *mgmtnetwork.Subnet) {
+	if subnet.ServiceEndpoints == nil {
+		subnet.ServiceEndpoints = &[]mgmtnetwork.ServiceEndpointPropertiesFormat{}
+	}
+
+	serviceEndpoint := mgmtnetwork.ServiceEndpointPropertiesFormat{
+		Service:   to.StringPtr(endpoint),
+		Locations: &[]string{"*"},
+	}
+
+	*subnet.ServiceEndpoints = append(*subnet.ServiceEndpoints, serviceEndpoint)
+}
```

Comment:
why are these pointer functions ? They don't reference the object.

Reply:
Thanks for the review! 

I added them (function receivers) because I come from a more strict  OOP background and I wanted to "encapsulate" all that behavior inside the same struct.

But as you mention is not necessary, I have just removed them.  

Tks. 🙂

Reply:
They could've remained receiver functions, but they did not need to be pointer receiver functions. They could've been a normal receiver function that copies the struct by value instead of reference.

Reply:
they could have, but should they have been ? we don't use the receiver fields/functions even once. it seems unnecessary to have a copy of an object we don't use